src/uct/tcp/tcp_ep.c

```diff
@@ -463,26 +552,42 @@ ucs_status_t uct_tcp_ep_am_short(uct_ep_h uct_ep, uint8_t am_id, uint64_t header
     uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
     uct_tcp_am_hdr_t *hdr;
 
-    UCT_TCP_AM_PREPARE(ep, am_id, iface->config.short_size - sizeof(*hdr),
-                       hdr, memcpy, payload, length, header, SHORT, "am_short");
+    UCT_CHECK_LENGTH(length + sizeof(header), 0,
+                     iface->config.short_size - sizeof(uct_tcp_am_hdr_t),
+                     "am_short");
+
+    UCT_TCP_AM_PREPARE(iface, ep, am_id, hdr, memcpy,
+                       payload, length, header, SHORT);
 
     uct_tcp_ep_am_send(iface, ep, hdr);
+    UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, length + sizeof(header));
     return UCS_OK;
 }
 
 ssize_t uct_tcp_ep_am_bcopy(uct_ep_h uct_ep, uint8_t am_id,
                             uct_pack_callback_t pack_cb, void *arg,
                             unsigned flags)
 {
-    uct_tcp_ep_t *ep = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_ep_t *ep       = ucs_derived_of(uct_ep, uct_tcp_ep_t);
     uct_tcp_iface_t *iface = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    ssize_t payload_length;
     uct_tcp_am_hdr_t *hdr;
 
-    UCT_TCP_AM_PREPARE(ep, am_id, iface->config.buf_size - sizeof(*hdr),
-                       hdr, pack_cb, arg, NULL, NULL, BCOPY, "am_bcopy");
+    UCT_TCP_AM_PREPARE(iface, ep, am_id, hdr, pack_cb,
+                       arg, NULL, NULL, BCOPY);
+
+    /* Save the length of the payload, because hdr (ep::buf)
+     * can be released inside `uct_tcp_ep_am_send` call */
+    payload_length = hdr->length;
+
+    UCT_CHECK_LENGTH_EX(payload_length, 0,
```

Comment:
what's the point to check length here? we already overrode the descriptor memory..

Reply:
yes, I know. But it looks like we are unable to check the length in BCOPY.